Azure: Delete-ResourceGroup lock fix

adriansuhov opened the pull request:
Until now, if -UseExistingRG was set, the framework would not check if the existing RG had a lock on it. As a result, the function Delete-ResourceGroup would enter an infinite loop. Added a check for lock, and if any lock is set on the VM, the RG cleanup will be skipped and a warn would show up in the logs. 
- Also added RG cleanup becasue it was missing. Only the resources inside the RG were cleaned with the old code.
### Before this commit, infinite loop
.............
03/06/2019 12:35:20 : [ERROR] Failed to delete resource PrimaryNIC-LISAv2-OneVM-ubuntuAz-SC48-20190227112013-role-0. We will try to remove it in next attempt.
03/06/2019 12:35:20 : [INFO ] Removing resource PublicIPv4-89767, type Microsoft.Network/publicIPAddresses
03/06/2019 12:35:20 : [ERROR] Failed to delete resource PublicIPv4-89767. We will try to remove it in next attempt.
03/06/2019 12:35:20 : [INFO ] Removing resource VirtualNetwork, type Microsoft.Network/virtualNetworks
03/06/2019 12:35:20 : [ERROR] Failed to delete resource VirtualNetwork. We will try to remove it in next attempt.
03/06/2019 12:35:20 : [INFO ] Removing resource AvailibilitySet, type Microsoft.Compute/availabilitySets
03/06/2019 12:35:20 : [ERROR] Failed to delete resource AvailibilitySet. We will try to remove it in next attempt.
03/06/2019 12:35:20 : [INFO ] Removing resource LISAv2-OneVM-ubuntuAz-SC48-20190227112013-role-0, type Microsoft.Compute/virtualMachines
03/06/2019 12:35:20 : [ERROR] Failed to delete resource LISAv2-OneVM-ubuntuAz-SC48-20190227112013-role-0. We will try to remove it in next attempt.
03/06/2019 12:35:20 : [INFO ] Removing resource LoadBalancer, type Microsoft.Network/loadBalancers
03/06/2019 12:35:20 : [ERROR] Failed to delete resource LoadBalancer. We will try to remove it in next attempt.
03/06/2019 12:35:20 : [INFO ] Removing resource PrimaryNIC-LISAv2-OneVM-ubuntuAz-SC48-20190227112013-role-0, type Microsoft.Network/networkInterfaces
03/06/2019 12:35:21 : [ERROR] Failed to delete resource PrimaryNIC-LISAv2-OneVM-ubuntuAz-SC48-20190227112013-role-0. We will try to remove it in next attempt.
03/06/2019 12:35:21 : [INFO ] Removing resource PublicIPv4-89767, type Microsoft.Network/publicIPAddresses
...........

### After this commit
03/06/2019 12:56:04 : [INFO ] Checking if LISAv2-OneVM-adsuhoTest-HD16-20190215143245 exists...
**03/06/2019 12:56:05 : [WARN ] Lock is in place for LISAv2-OneVM-adsuhoTest-HD16-20190215143245. Skipping RG delete!**
03/06/2019 12:56:05 : [INFO ] Failed to delete resource group LISAv2-OneVM-adsuhoTest-HD16-20190215143245.. Please delete it manually.
03/06/2019 12:56:05 : [INFO ] Checking background cleanup jobs..
03/06/2019 12:56:05 : [INFO ] All background cleanup jobs finished.
03/06/2019 12:56:05 : [INFO ] Removed 0 context files.
03/06/2019 12:56:05 : [INFO ] Test XH59 finished

**RG removal addition**
03/06/2019 13:15:18 : [INFO ] Resources in LISAv2-OneVM-ubuntuAz-SC48-20190227112013 are all deleted.
**VERBOSE: Performing the operation "Removing resource group ..." on target "LISAv2-OneVM-ubuntuAz-SC48-20190227112013".**
03/06/2019 13:16:03 : [INFO ] Successfully delete resource group LISAv2-OneVM-ubuntuAz-SC48-20190227112013..

<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

Issue # <!-- if any -->
<!--
Please add an informative description that covers that changes made by the pull request.
This checklist is used to make sure that common issues in a pull request are addressed.
This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.
-->

---

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [x] Verify PR checker results and fix any coding errors.
- [x] Included LISAv2 sample test results to demonstrate code functionality.

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/LIS/LISAv2/blob/master/.github/CONTRIBUTING.md).

iamshital commented on Libraries/Azure.psm1:
```diff
@@ -362,21 +362,28 @@ Function Delete-ResourceGroup([string]$RGName, [switch]$KeepDisks, [bool]$UseExi
-                    catch {
-                        Write-LogErr "Failed to delete resource $($resource.Name). We will try to remove it in next attempt."
+            # Get RG lock. If there is any lock in place, don't try to delete the Resource Group
+            $rmLock = Get-AzureRmResourceLock -ResourceGroupName $RGName
```
Lock could be given to ResourceGroup (parent lock) or individual resources (child lock).
My suggestion -
1. If lock is given to ResourceGroup, then skip cleanup of all inside resources ( your current implementation )
2. If lock is given to Child Resource, then you can continue cleaning up other resources, keeping only Locked resource. This will also skip resource group cleanup. This scenario is particularly helpful when we pin "AvailibilitySet" (or any other resource type) to test cluster, and use existing resource group for deployment. This way, LISAv2 will continue deploy/cleanup in existing resource group, without cleaning "AvailabilitySet".

adriansuhov replied:
Sounds good. Will make the changes tomorrow. Thanks for the suggestion!

lizzha commented on Libraries/Azure.psm1:
```diff
@@ -362,21 +362,28 @@ Function Delete-ResourceGroup([string]$RGName, [switch]$KeepDisks, [bool]$UseExi
                 }
-                $CurrentResources = Get-AzureRmResource -ResourceGroupName $RGName
+                Write-LogInfo "Resources in $RGName are all deleted."
+                Remove-AzureRmResourceGroup -Name $RGName -Verbose -Force
```
it's the design that when using existing resource group, do not remove the RG when clean up, as the RG is not created by automation.

adriansuhov replied:
Ok. I'll redo this change

adriansuhov commented:
@lizzha I've reverted the RG removal
@iamshital I added the requested feature. 

Logs for partial resource deletion (removed verbose logs): 
03/07/2019 12:19:16 : [INFO ] Try to delete resource group LISAv2-OneVM-adsuhoTest-HD16-20190215143245...
03/07/2019 12:19:16 : [INFO ] Checking if LISAv2-OneVM-adsuhoTest-HD16-20190215143245 exists...
Removing resource LISAv2-OneVM-adsuhoTest-HD16-20190215143245-role-0, type Microsoft.Compute/virtualMachines
Removing resource LISAv2-OneVM-adsuhoTest-HD16-20190215143245-role-0/CustomScript, type Microsoft.Compute/virtualMachines/extensions
Removing resource PrimaryNIC-LISAv2-OneVM-adsuhoTest-HD16-20190215143245-role-0, type Microsoft.Network/networkInterfaces
Removing resource PrimaryNIC-LISAv2-OneVM-adsuhoTest-HD16-20190215143245-role-0, type Microsoft.Network/networkInterfaces
03/07/2019 12:21:07 : [INFO ] Resources in LISAv2-OneVM-adsuhoTest-HD16-20190215143245 are deleted.
